File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
import sys
input = sys.stdin.readline
n,m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
prefix = [[0] * (n+1) for _ in range(n+1)]
# 왜 y first ?
for x in range(n):
    for y in range(n):
        prefix[x+1][y+1] = prefix[x+1][y] + prefix[x][y+1] - prefix[x][y] + arr[x][y]

for t in range(m):
    x1, y1, x2, y2 = map(int, input().split())
    ans = prefix[x2][y2] + prefix[x1-1][y1-1] - prefix[x2][y1-1] - prefix[x1-1][y2]
    print(ans)

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.Scanner;

public class Main {

  public static void main(String[] args) throws IOException {
    Scanner sc = new Scanner(System.in);
    int N = sc.nextInt();
    int M = sc.nextInt();
    int[][] arr = new int[N+1][N+1];
    for (int i = 1; i <= N; i++) {
      for (int j = 1; j <= N; j++) {
        arr[i][j] = sc.nextInt();
      }
    }

    // prefix
    int[][] prefix = new int[N+1][N+1];
    prefix[1][1] = arr[1][1];
    for (int i = 1; i <= N; i++) {
      for (int j = 1; j <= N; j++) {
        prefix[i][j] = prefix[i][j-1] + prefix[i-1][j] - prefix[i-1][j-1]+ arr[i][j];
      }
    }
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    while (M-- > 0) {
      int x1 = sc.nextInt();
      int y1 = sc.nextInt();
      int x2 = sc.nextInt();
      int y2 = sc.nextInt();

      int ans = 0;
      ans = prefix[x2][y2] - (prefix[x1-1][y2] + prefix[x2][y1-1]) + prefix[x1-1][y1-1];
      bw.write(ans + "\n");
    }
    bw.flush();
  }
}
